File: src/Solutions/Euler120.hs
module Solutions.Euler120 where

-- See problem 123 for explanation.
remainder a n = if even n then 2 else 2 * a * n `mod` (a * a)
maxRemainder n = foldr (max . remainder n) 0 [0..2 * n]

euler120 :: IO String
euler120 = return $ show $ sum $ map maxRemainder [3..1000]

File: rust/src/solutions/euler120.rs
// See problem 123 for explanation.

fn remainder(a: i32, n: i32) -> i32 {
    if n % 2 == 0 {
        2
    } else {
        (2 * a * n) % (a * a)
    }
}

fn max_remainder(n: i32) -> i32 {
    (0..=n + n).fold(0, |s, x| std::cmp::max(s, remainder(n, x)))
}

pub fn euler120() -> String {
    (3..=1000).map(max_remainder).sum::<i32>().to_string()
}
